### PR TITLE
Dashed indent guide lines and gutter for line numbers

### PR DESCRIPTION
**Pull request by Lessica**
## Major changes in this pull request

1. Use system monospaced font by default (A code editor uses monospaced fonts in most cases).
2. (Customizable) Draw a gutter for line number view.
3. Draw **dashed** indent guide lines by columns, not logical lines, which also support vertical layout.
4. Wrap some UI components in Window Pane with NSStackView.
5. Set minimum window size to `(400,400)` in XIB

Localization for languages other than Simplified Chinese were not updated.

## Screenshots

截屏2021-06-05 上午12 27 04

截屏2021-06-05 上午12 28 01

截屏2021-06-05 上午12 30 08

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,138 @@ default.profraw
...
+
+### Xcode Patch ###
+*.xcodeproj/*
+!*.xcodeproj/project.pbxproj
+!*.xcodeproj/xcshareddata/
+!*.xcworkspace/contents.xcworkspacedata
+**/xcshareddata/WorkspaceSettings.xcsettings
+
+# End of https://www.toptal.com/developers/gitignore/api/macos,xcode,swift
+
```
Comment (Lessica) [inline .gitignore:150]: These lines are not necessary to be merged.

File: CotEditor/Sources/DefaultSettings.swift [hunk 2]
```diff
@@ -61,7 +62,7 @@ struct DefaultSettings {
         .splitViewVertical: false,
         .windowAlpha: 1.0,
         
-        .fontName: (NSFont.userFont(ofSize: 0) ?? NSFont.systemFont(ofSize: 0)).fontName,
+        .fontName: (NSFont.monospacedSystemFont(ofSize: 0, weight: .regular)).fontName,
```
Comment (Lessica) [inline CotEditor/Sources/DefaultSettings.swift:65]: I think it'll be better to use monospaced fonts in a code editor app.

File: CotEditor/Sources/EditorTextView.swift [hunk 3]
```diff
@@ -984,7 +984,7 @@ final class EditorTextView: NSTextView, Themable, CurrentLineHighlighting, URLDe
             
             let guideRect = NSRect(x: isRTL ? self.bounds.width - x : x,
                                    y: dirtyRect.minY,
-                                   width: 1.0,
+                                   width: 0.5,
```
Comment (Lessica) [inline CotEditor/Sources/EditorTextView.swift:987]: A smaller guide line width would be better.

File: CotEditor/Sources/LayoutManager.swift [hunk 4]
```diff
@@ -302,6 +302,9 @@ extension LayoutManager: NSLayoutManagerDelegate {
 
 private extension NSLayoutManager {
     
+    typealias IndentGuideLine = (lineIndex: Int, lineRange: NSRange, indexes: [Int])
+    typealias IndentGuideUnit = (lineIndex: Int, columnIndex: Int, linePoints: [CGPoint])
+
```
Comment (Lessica) [inline CotEditor/Sources/LayoutManager.swift:307]: Maybe we need two more `struct`.

File: CotEditor/Base.lproj/WindowPane.storyboard [hunk 5]
```diff
@@ -49,135 +49,214 @@
...
+                                            <integer value="1000"/>
+                                            <integer value="1000"/>
+                                            <integer value="1000"/>
+                                        </visibilityPriorities>
+                                        <customSpacing>
+                                            <real value="3.4028234663852886e+38"/>
+                                            <real value="3.4028234663852886e+38"/>
+                                            <real value="3.4028234663852886e+38"/>
+                                        </customSpacing>
+                                    </stackView>
```
Comment (Lessica) [inline CotEditor/Base.lproj/WindowPane.storyboard:208]: I prefer NSStackView here.

File: CotEditor/Sources/DocumentViewController.swift [hunk 6]
```diff
@@ -156,6 +158,7 @@ final class DocumentViewController: NSSplitViewController, ThemeHolder, NSTextSt
         
         return super.restorableStateKeyPaths + [
             #keyPath(showsLineNumber),
+            #keyPath(showsGutterForLineNumber),
```
Comment (Lessica) [inline CotEditor/Sources/DocumentViewController.swift:161]: I wonder why these properties are restorable.

File: CotEditor/zh-Hans.lproj/WindowPane.strings [hunk 7]
```diff
@@ -43,11 +43,13 @@
 "iw9-Ba-eD2.title" = "显示：";
 /* Class = "NSButtonCell"; title = "Line numbers"; ObjectID = "A9U-Sx-or3"; */
 "A9U-Sx-or3.title" = "行号";
+/* Class = "NSButtonCell"; title = "Gutter"; ObjectID = "uY8-8c-yDw"; */
+"uY8-8c-yDw.title" = "分隔线";
```
Comment (Lessica) [inline CotEditor/zh-Hans.lproj/WindowPane.strings:47]: How to update localization automatically?

**Comment by Lessica** [comment]
I want to be a member of CotEditor team.
We can work together to make better native macOS Apps.

My app: [JSTColorPicker](https://82flex.com/jstcpweb/)

**Comment by 1024jp** [comment]
Please make pull-requests topic by topic. Some of your requests are understandable and have worth to think but others are not. In addition, it's better to make a ticket first and discuss it with the maintainer when proposing a new feature. 

Well, I comment on your requests:

> 1. Use system monospaced font by default (A code editor uses monospaced fonts in most cases).

First, CotEditor is not a code editor but a general plain-text editor that a non-negligible number of users use to write daily text. So, I dare to set a proportional font. I understand many of the users use it for coding and change the font setting first to monospace font. I also have a future improvement idea for this, although not implemented yet. So, your request at this time point in this way is not acceptable.

> 2. (Customizable) Draw a gutter for line number view.

I am still considering whether I should add this option. I want to be extremely careful when I add an option in preference. Adding an option is easy, but then the setting UI becomes easily messy. I understand some of the users miss the border beside the line number view.
Furthermore, the code for this option is already finished locally. So I don't need a pull request for this even if I decided to add the option.

> 3. Draw dashed indent guide lines by columns, not logical lines, which also support vertical layout.

Sorry, I did not understand this change. You just changed the line style to dashed? But why?

> 4. Wrap some UI components in Window Pane with NSStackView.

I don't think this change is required.

> 5. Set minimum window size to (400,400) in XIB

Not acceptable. I dare not set the minimum window size for uses who want to use small size windows.

**Comment by 1024jp** [comment]
> Well, I want to be a member of CotEditor team. We can work together to make better native macOS Apps.

Thank you for your interest in the CotEditor project. All your suggested next features are important. I agree with you (I am personally not a fan of minimap though).

However, I have my own roadmap and (strong) design philosopy for CotEditor, and therefore, if we really collaborate, we need to discuss a lot first on the design (I mean both implementation design in code and aesthetic design in UI), and then can share the implementation tasks.
At this time point, I don't have a plan to assign new committers who can change CotEditor freely.

So if you have a specific design suggestion, make an issue ticket or a draft pull-request, and let us discuss it.

**Comment by Lessica** [comment]
> > 3. Draw dashed indent guide lines by columns, not logical lines, which also support vertical layout.
> 
> Sorry, I did not understand this change. You just changed the line style to dashed? But why?

Drawing line instead of rectangle for each glyph makes indent guide lines more customizable. We can apply more styles to these lines (`dashed` is just one option).

**Comment by Lessica** [comment]
> > 4. Wrap some UI components in Window Pane with NSStackView.
> 
> I don't think this change is required.

Constraints in that XIB are in a mess. I hope we can group stacked UI components such as checkbox, radio.

**Comment by Lessica** [comment]
> > 5. Set minimum window size to (400,400) in XIB
> 
> Not acceptable. I dare not set the minimum window size for uses who want to use small size windows.

I think a **functional** minimum window size is required.

**Comment by Lessica** [comment]
> > Well, I want to be a member of CotEditor team. We can work together to make better native macOS Apps.
> 
> Thank you for your interest in the CotEditor project. All your suggested next features are important. I agree with you (I am personally not a fan of minimap though).
> 
> However, I have my own roadmap and (strong) design philosopy for CotEditor, and therefore, if we really collaborate, we need to discuss a lot first on the design (I mean both implementation design in code and aesthetic design in UI), and then can share the implementation tasks.
> At this time point, I don't have a plan to assign new committers who can change CotEditor freely.
> 
> So if you have a specific design suggestion, make an issue ticket or a draft pull-request, and let us discuss it.

I am also interested in TextKit developing. If you want a partner later, please contact me.

Thank you for your reply.

**Comment by Lessica** [comment]
> > 1. Use system monospaced font by default (A code editor uses monospaced fonts in most cases).
> 
> First, CotEditor is not a code editor but a general plain-text editor that a non-negligible number of users use to write daily text. So, I dare to set a proportional font. I understand many of the users use it for coding and change the font setting first to monospace font. I also have a future improvement idea for this, although not implemented yet. So, your request at this time point in this way is not acceptable.

OK, understand.👌

**Comment by Lessica** [comment]
> > 2. (Customizable) Draw a gutter for line number view.
> 
> I am still considering whether I should add this option. I want to be extremely careful when I add an option in preference. Adding an option is easy, but then the setting UI becomes easily messy. I understand some of the users miss the border beside the line number view.
> Furthermore, the code for this option is already finished locally. So I don't need a pull request for this even if I decided to add the option.

The thing I missed is not just a border beside the line number view, it's a code folding ribbon 🤣.

**Pull request closed without merging**